File: src/omnia/data/structure/tuple/Couple.java
package omnia.data.structure.tuple;

import java.util.function.Function;

public interface Couple<A, B> extends Tuples.AtLeastCouple<A, B>, Tuples.AtMostCouple {

  @Override
  default int count() {
    return 2;
  }

  @Override
  <R> Couple<R, B> mapFirst(Function<? super A, ? extends R> mapper);

  @Override
  <R> Couple<A, R> mapSecond(Function<? super B, ? extends R> mapper);

  @Override
  <T> Triple<A, B, T> append(T object);

  @Override
  <C, D> Quadruple<A, B, C, D> append(Couple<C, D> other);

  @Override
  <C, D, E> Quintuple<A, B, C, D, E> append(Triple<C, D, E> other);

  @Override
  <C, D, E, F> Sextuple<A, B, C, D, E, F> append(Quadruple<C, D, E, F> other);

  @Override
  <C, D, E, F, G> Septuple<A, B, C, D, E, F, G> append(Quintuple<C, D, E, F, G> other);

  @Override
  <C, D, E, F, G, H> Octuple<A, B, C, D, E, F, G, H> append(Sextuple<C, D, E, F, G, H> other);

  @Override
  <C, D, E, F, G, H, I> Nonuple<A, B, C, D, E, F, G, H, I> append(Septuple<C, D, E, F, G, H, I> other);

  @Override
  <C, D, E, F, G, H, I, J> Decuple<A, B, C, D, E, F, G, H, I, J> append(Octuple<C, D, E, F, G, H, I, J> other);
}

File: src/omnia/data/structure/tuple/Couple.kt
package omnia.data.structure.tuple

import omnia.data.structure.tuple.Tuples.AtLeastCouple
import omnia.data.structure.tuple.Tuples.AtMostCouple

interface Couple<A, B> : AtLeastCouple<A, B>, AtMostCouple {

  override fun count(): Int {
    return 2
  }

  override fun <R> mapFirst(mapper: (A) -> R): Couple<R, B>
  override fun <R> mapSecond(mapper: (B) -> R): Couple<A, R>
  override fun <T> append(`object`: T): Triple<A, B, T>
  override fun <C, D> append(other: Couple<C, D>): Quadruple<A, B, C, D>
  override fun <C, D, E> append(other: Triple<C, D, E>): Quintuple<A, B, C, D, E>
  override fun <C, D, E, F> append(other: Quadruple<C, D, E, F>): Sextuple<A, B, C, D, E, F>
  override fun <C, D, E, F, G> append(other: Quintuple<C, D, E, F, G>): Septuple<A, B, C, D, E, F, G>
  override fun <C, D, E, F, G, H> append(other: Sextuple<C, D, E, F, G, H>): Octuple<A, B, C, D, E, F, G, H>
  override fun <C, D, E, F, G, H, I> append(other: Septuple<C, D, E, F, G, H, I>): Nonuple<A, B, C, D, E, F, G, H, I>
  override fun <C, D, E, F, G, H, I, J> append(other: Octuple<C, D, E, F, G, H, I, J>): Decuple<A, B, C, D, E, F, G, H, I, J>
}